language/bytecode-verifier/src/verifier.rs

```diff
@@ -53,23 +53,25 @@ pub fn verify_main_signature(script: &CompiledScript) -> VMResult<()> {
 }
 
 fn verify_main_signature_impl(script: &CompiledScript) -> PartialVMResult<()> {
-    fn is_valid_arg_type(idx: usize, arg_type: &SignatureToken) -> bool {
-        use SignatureToken as S;
-        match arg_type {
-            // &signer is a type that can only be populated by the Move VM. And its value is filled
-            // based on the sender of the transaction
-            S::Reference(inner) => idx == 0 && matches!(&**inner, S::Signer),
-            _ => arg_type.is_valid_for_constant(),
-        }
+    use SignatureToken as S;
+    let arguments = &script.signature_at(script.as_inner().parameters).0;
+    // Remove the prefix of signer types
```

Comment:
Nit: comment slightly confusing because this isn't actually removing anything. Maybe something like "check that all `&signer` formals occur before non-`&signer` formals"?